lib/xandra/connection.ex

```diff
@@ -509,11 +513,32 @@ defmodule Xandra.Connection do
   end
 
   def disconnected(:cast, {:release_stream_id, stream_id}, %__MODULE__{} = data) do
-    data = update_in(data.free_stream_ids, &MapSet.put(&1, stream_id))
     data = update_in(data.in_flight_requests, &Map.delete(&1, stream_id))
     {:keep_state, data}
   end
 
+  def disconnected(:cast, {:timed_out_id, stream_id}, %__MODULE__{} = data) do
+    data = update_in(data.in_flight_requests, &Map.delete(&1, stream_id))
+    data = update_in(data.timed_out_ids, &MapSet.put(&1, stream_id))
+
+    actions = [
+      {{:timeout, {:stream_id, stream_id}}, @restore_timed_out_stream_id_timeout,
```

Comment:
We should try to avoid setting potentially-thousands of timeouts here. Instead, what we could do is store the timestamp that a stream ID timed out at and then periodically clean those timeouts. For example, stored `timed_out_ids` as `%{id => timed_out_at, ...}`. Then, every 30 seconds, flush the ones that are older than 5 minutes. Makes sense?

Reply:
Agreed